File: BackEnd/PHP/laravelshop/app/Http/Controllers/Api/V1/Crawler/CrawlerController.php
<?php

namespace App\Http\Controllers\Api\V1\Crawler;

use OpenApi\Attributes as OA;

use App\Http\Controllers\Api\V1\ApiController;
use App\Modules\Crawler\Requests\StoreCrawlerJobRequest;
use App\Modules\Crawler\Requests\TriggerCrawlerRequest;
use App\Modules\Crawler\Requests\UpdateCrawlerJobRequest;
use App\Modules\Crawler\Services\CrawlerService;
use Illuminate\Http\Request;

#[OA\Tag(name: 'Crawler')]
class CrawlerController extends ApiController
{
    public function __construct(private readonly CrawlerService $service) {}

    public function jobs(Request $request)
    {
        return $this->success($this->service->jobs($request->user(), $request->query('status'), (int) $request->query('page', 1), (int) $request->query('limit', 20)));
    }

    public function storeJob(StoreCrawlerJobRequest $request)
    {
        return $this->success($this->service->createJob($request->user(), $request->validated()), [], 201);
    }

    public function updateJob(UpdateCrawlerJobRequest $request, int $id)
    {
        return $this->success($this->service->updateJob($request->user(), $id, $request->validated()));
    }

    public function destroyJob(Request $request, int $id)
    {
        return $this->success($this->service->deleteJob($request->user(), $id));
    }

    public function runJob(Request $request, int $id)
    {
        return $this->success($this->service->runJob($request->user(), $id));
    }

    public function trigger(TriggerCrawlerRequest $request)
    {
        return $this->success($this->service->trigger($request->user(), $request->validated()));
    }

    public function runs(Request $request)
    {
        return $this->success($this->service->runs($request->user(), $request->query('status'), $request->query('jobId') ? (int) $request->query('jobId') : null, (int) $request->query('page', 1), (int) $request->query('limit', 20)));
    }

    public function monitoring(Request $request)
    {
        return $this->success($this->service->monitoring($request->user()));
    }
}

File: BackEnd/Java/springshop_maven/src/main/java/com/pbshop/springshop/crawler/CrawlerController.java
package com.pbshop.springshop.crawler;

import java.util.Map;

import jakarta.validation.Valid;

import org.springframework.http.HttpStatus;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.ResponseStatus;
import org.springframework.web.bind.annotation.RestController;

import com.pbshop.springshop.auth.security.AuthenticatedUserPrincipal;
import com.pbshop.springshop.common.api.ApiResponse;
import com.pbshop.springshop.crawler.dto.CrawlerDtos.CreateCrawlerJobRequest;
import com.pbshop.springshop.crawler.dto.CrawlerDtos.TriggerCrawlerRequest;
import com.pbshop.springshop.crawler.dto.CrawlerDtos.UpdateCrawlerJobRequest;

@RestController
@RequestMapping("${pbshop.api.base-path:/api/v1}/crawler/admin")
public class CrawlerController {

    private final CrawlerService crawlerService;

    public CrawlerController(CrawlerService crawlerService) {
        this.crawlerService = crawlerService;
    }

    @GetMapping("/jobs")
    public ApiResponse<Map<String, Object>> jobs(
            @AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @RequestParam(required = false) String status,
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "20") int limit
    ) {
        return ApiResponse.success(crawlerService.jobs(principal, status, page, limit));
    }

    @PostMapping("/jobs")
    @ResponseStatus(HttpStatus.CREATED)
    public ApiResponse<Map<String, Object>> createJob(
            @AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @Valid @RequestBody CreateCrawlerJobRequest request
    ) {
        return ApiResponse.success(crawlerService.createJob(principal, request));
    }

    @PatchMapping("/jobs/{id}")
    public ApiResponse<Map<String, Object>> updateJob(
            @AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @PathVariable Long id,
            @RequestBody UpdateCrawlerJobRequest request
    ) {
        return ApiResponse.success(crawlerService.updateJob(principal, id, request));
    }

    @DeleteMapping("/jobs/{id}")
    public ApiResponse<Map<String, Object>> deleteJob(
            @AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @PathVariable Long id
    ) {
        return ApiResponse.success(crawlerService.deleteJob(principal, id));
    }

    @PostMapping("/jobs/{id}/run")
    public ApiResponse<Map<String, Object>> runJob(
            @AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @PathVariable Long id
    ) {
        return ApiResponse.success(crawlerService.runJob(principal, id));
    }

    @PostMapping("/triggers")
    public ApiResponse<Map<String, Object>> trigger(
            @AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @RequestBody TriggerCrawlerRequest request
    ) {
        return ApiResponse.success(crawlerService.trigger(principal, request));
    }

    @GetMapping("/runs")
    public ApiResponse<Map<String, Object>> runs(
            @AuthenticationPrincipal AuthenticatedUserPrincipal principal,
            @RequestParam(required = false) String status,
            @RequestParam(required = false) Long jobId,
            @RequestParam(defaultValue = "1") int page,
            @RequestParam(defaultValue = "20") int limit
    ) {
        return ApiResponse.success(crawlerService.runs(principal, status, jobId, page, limit));
    }

    @GetMapping("/monitoring")
    public ApiResponse<Map<String, Object>> monitoring(@AuthenticationPrincipal AuthenticatedUserPrincipal principal) {
        return ApiResponse.success(crawlerService.monitoring(principal));
    }
}
